SelfAndPeerEvaluations/partner_eval_quiz_final_grader.py

```diff
@@ -0,0 +1,101 @@
+import canvasapi
+from collections import defaultdict
+import datetime
+import json
+import os
+from utils import *
+
+
+class PartnerEvalQuizGrader:
+    def __init__(
+        self,
+        course: canvasapi.course.Course,
+        json_questions: dict,
+        assignment_group: canvasapi.assignment.AssignmentGroup = None,
+    ):
+        self.course = course
+        self.json_questions = json_questions
+
+        self.assignment_group = (
+            assignment_group
+            if assignment_group is not None
+            else find_partner_eval_ag(course)
+        )
+        # make sure that assignment group has assignment attribute
+        if not hasattr(self.assignment_group, "assignments"):
+            self.assignment_group = course.get_assignment_group(
+                self.assignment_group.id, include=["assignments"]
+            )
+
+        self.student_id_map = make_student_id_map(self.course)
+
+    def upload_final_grades_to_canvas(
+        self, assignment_name="Overall Evaluation Grade"
+    ) -> None:
+        files = []
+        print("We'll be basing final grade off of these files:")
+        print("These files will be processed in proper sorted order.")
+        for filename in os.listdir("./quiz_results/quiz_reports"):
+            print("   ", filename)
+            files.append(os.path.join("./quiz_results/quiz_reports/", filename))
+
+        individual_students_stats = []
+        csv_files = {}
+        for name, id in self.student_id_map.items():
+            individual_students_stats.append(
+                PartnerEvalQuizIndividualStats(name, id, sorted(files))
+            )
+            csv_files[id] = individual_students_stats[-1].csv_file_path
+
+        print("ok i made the csvs")
+        assignment = self.course.create_assignment(
+            assignment={
+                "name": assignment_name,
+                "description": "Your overall group work score based on the past Partner Evaluation assignments.",
+                "assignment_group_id": self.assignment_group.id,
+                "submission_types": ["none"],
+                "points_possible": 100,
+                "due_at": datetime.datetime.now().isoformat(),
+                "lock_at": datetime.datetime.now().isoformat(),
+                "unlock_at": datetime.datetime.now().isoformat(),
+                "published": True,
+            }
+        )
+        print("Uploading final grades and csv files to Canvas")
+        print("This might take a while")
+        assignment.submissions_bulk_update(
```

Comment:
`submissions_bulk_update`  returns a progress object which you can use to check the status of the bulk_update to see when it has been completed.
```suggestion
        upload_progress = assignment.submissions_bulk_update(
```